### PR TITLE
#165020217 User functionality 

### PR DESCRIPTION
**Pull request by fridolinho**
## What does this PR do?
User functionality
## Description of Task to be completed?
- Authenticate routes using jwt
- retrieve a list of existing authors using a PG query to the database
- check if username exists and retrieve author profile details with that username
## How should this be manually tested?

1. Run these commands
     - clone https://github.com/andela/ah-kgl-avengers-backend.git
     - git checkout ft-user-functionality-165020217
     - npm install
     - npm test
2. [check travis build here](https://travis-ci.com/andela/ah-kgl-avengers-backend/builds/108810973)
3. Postman
    - Send a GET request to localhost:3000/api/v1/profiles/:username  
       create a user and use his username in the url to see his profile.
    - Send a GET request to localhost:3000/api/v1/users/:authors to get a list of all authors registered

## What are the relevant pivotal tracker stories?
[user functionality #165020217](https://www.pivotaltracker.com/story/show/165020217)
## Screenshots (if appropriate)
image

## Questions:
N/A

### CODE REVIEW

**Review by aaronsekisambu — CHANGES_REQUESTED** [review]
@fridolinho 
1. please make sure you are not pushing the .vscode file.
2. Also, cross-check if all your tests are passing in all scenarios
3. You are missing mock-ups for this feature
4. Write a descriptive message on how to test this feature manually. I do not know where to start apart from testing tests only.
5. Remember to squash your commits into one before your branch gets merged to develop

File: routes/api/users.js [hunk 1]
```diff
@@ -37,4 +37,17 @@ router.post(
   userControllers.logout
 );
 
+// User functionality
+router.get(
+  '/users/authors',
+  passport.authenticate('jwt', { session: false }),
```
Comment (avpaul) [inline routes/api/users.js:43]: I think authentication should be optional as we want to view author information only and this functionality should be available to all users even if they are not authenticated.
Check the [readme](https://github.com/andela/ah-kgl-avengers-backend#get-profile) file of the project.

File: routes/api/users.js [hunk 2]
```diff
@@ -37,4 +37,17 @@ router.post(
...
+  userControllers.getAllAuthors
+);
+router.get(
```
Comment (avpaul) [inline routes/api/users.js:46]: Same as above.

File: tests/user.test.js [hunk 3]
```diff
@@ -129,4 +129,28 @@ describe('User', () => {
         });
     });
   });
+
+  describe('user functionality', () => {
```
Comment (avpaul) [inline tests/user.test.js:133]: When rebasing or merging, remember to implement login(to get a token as we have got a working solution of getting a token). The test should work if you either make token verification optional or add a token.

**Review by avpaul — CHANGES_REQUESTED** [review]
Looks good just implement the feedback left.

File: controllers/users.js [hunk 4]
```diff
@@ -284,6 +284,42 @@ class Users {
...
+
+  /**
+   * get a list of authors
+   * @param {object} req
+   * @param {object} res
+   * @param {object} next
+   * @returns {object} res
+   */
+  static async getAllAuthors(req, res) {
+    const result = await User.findAll({ attributes: ['image', 'username', 'email'] });
```
Comment (niyobobo) [inline controllers/users.js:296]: For the long-running operation like this of dealing with a database, it's better if you putt them in error handler block

**Review by niyobobo — CHANGES_REQUESTED** [review]

File: routes/api/users.js (hunk 1 above)
Comment (daud1) [reply routes/api/users.js:43]: The ticket description requires that only authenticated/logged in users are able to see the list of existing authors and not just any user.

**Review by daud1 — CHANGES_REQUESTED** [review]
LGTM 👍 
However, your PR description is not very useful in guiding a reviewer.
Please fix that before I can merge it in.

File: controllers/users.js (hunk 4 above)
Comment (fridolinho) [reply controllers/users.js:296]: I got what you mean but i am not using try/catch to avoid overuse.

**Comment by fridolinho** [comment]
> LGTM 👍
> However, your PR description is not very useful in guiding a reviewer.
> Please fix that before I can merge it in.

Fixed

**Review by daud1 — APPROVED** [review]

**Pull request merged**